ddtrace/internal/coverage/code.py

```diff
@@ -0,0 +1,264 @@
+            return ctx_covered.get()
+        return self.covered
+
+    class CollectInContext:
+        def __enter__(self):
+            ctx_covered.set(defaultdict(set))
+            ctx_coverage_enabed.set(True)
+
+        def __exit__(self, *args, **kwargs):
+            ctx_coverage_enabed.set(False)
+
+    @classmethod
+    def start_coverage(cls):
+        if cls._instance is None:
+            return
+        cls._instance.coverage_enabled = True
+
+    @classmethod
+    def stop_coverage(cls):
+        if cls._instance is None:
+            return
+        cls._instance.coverage_enabled = False
+
+    @classmethod
+    def coverage_enabled(cls):
+        if ctx_coverage_enabed.get():
+            return True
+        if cls._instance is None:
+            return False
+        return cls._instance.coverage_enabled
+
+    @classmethod
+    def report_seen_lines(cls):
+        """Generate the same data as expected by ddtrace.ci_visibility.coverage.build_payload:
+
+        if input_path is provided, filter files to only include that path, and make it relative to said path
+
+        "files": [
+            {
+                "filename": <String>,
+                "segments": [
+                    [Int, Int, Int, Int, Int],  # noqa:F401
+                ]
+            },
+            ...
+        ]
+        """
+        if cls._instance is None:
+            return []
+        files = []
+        if ctx_coverage_enabed.get():
+            covered = ctx_covered.get()
+        else:
+            covered = cls._instance.covered
+        for path, lines in covered.items():
+            sorted_lines = sorted(lines)
+            collapsed_ranges = collapse_ranges(sorted_lines)
+            file_segments = []
+            for file_segment in collapsed_ranges:
+                file_segments.append([file_segment[0], 0, file_segment[1], 0, -1])
+            files.append({"filename": path, "segments": file_segments})
+
+        return files
+
+    def transform(self, code: CodeType, _module: ModuleType) -> CodeType:
+        code_path = Path(code.co_filename).resolve()
+        # TODO: Remove hardcoded paths
+        if not code_path.is_relative_to(CWD):
+            # Not a code object we want to instrument
+            return code
+
+        # Recursively instrument nested code objects, in topological order
+        # DEV: We need to make a list of the code objects because when we start
+        # mutating the parent code objects, the hashes maintained by the
+        # generator will be invalidated.
+        for nested_code, parent_code in list(collect_code_objects(code)):
+            # Instrument the code object
+            new_code = self.instrument_code(nested_code)
+
+            # If it has a parent, update the parent's co_consts to point to the
+            # new code object.
+            if parent_code is not None:
+                replace_in_tuple(parent_code.co_consts, nested_code, new_code)
```

Comment:
For anybody wondering why we do this hacky thing instead of recursively scanning the arguments of the abstract code instruction, looking for `CodeType` objects to replace, it turns out that the current approach is much faster 🤷